### ISSUE TITLE
can't reproduce the result of GANsynth

### ISSUE THREAD

**Issue by shansiliu95** [magenta/magenta#1732]
Hi, I have finished training with the recommended hyperparameters after 4 days. But the generated sound with the midi file is very poor and lacks diversity. I trained only on the Acoustic Subset and used tensorflow 1.15.2. How should I reproduce the results in the GANsynth paper? @jesseengel

**Comment by shansiliu95** [comment]
I used this command: python magenta/models/gansynth/gansynth_train.py --config=mel_prog_hires --hparams='{"train_data_path":"/path/to/nsynth-train.tfrecord", "train_root_dir":"/tmp/gansynth/train"}'

**Comment by shansiliu95** [comment]
This is the generated sound: https://drive.google.com/file/d/10K_DMILAOegQEq7i3kti7YbtMpAALvbL/view?usp=sharing

**Comment by jesseengel** [comment]
Interesting, sorry it's been a while since I trained one of these models, but at least I'm sure we checked the open source code reproduced results when we released. Is that example interpolating in z space? It sounds like mode collapse if that's the case, which was common earlier in training. Did you use a v100 to train? How many iterations did you get through? I'm curious if maybe it just needs to train longer.

**Comment by shansiliu95** [comment]
Thanks for your quick reply. Yes, I have a 32GB V100. Here are the training curves.
![image](https://user-images.githubusercontent.com/44488551/83065574-92285f00-a018-11ea-9b5d-8d3f79430bde.png)

**Comment by shansiliu95** [comment]
![image](https://user-images.githubusercontent.com/44488551/83065621-a2d8d500-a018-11ea-994c-0e84197ef845.png)

**Comment by shansiliu95** [comment]
I noticed other people had the same problem: https://github.com/tensorflow/magenta/issues/1607

**Comment by shansiliu95** [comment]
I am pretty sure something is wrong. My colleague trained another model with the recommended hyperparameter, and the generated music is similar to the music generated by me.

**Comment by shansiliu95** [comment]
I start re-training with a lower TensorFlow version: 1.13.1.

**Comment by jesseengel** [comment]
I agree, that's very strange. I found a model that I trained on GCP Nov 1, 2019, using the github code (mel_prog_hires, v100x1, 5 days, TF 1.14.0).

I generated from it and [got expected behavior](https://drive.google.com/file/d/1xqTWok_csr_jKNdsqhSiMt46ZHQoRPtn/view?usp=sharing). Can't pull up the training graphs atm, but I think what you have looks similar to what I remember.

My guess is either something funky happened between 1.14 and 1.15 or it's a problem with dataset mismatch (more likely). 

This model was trained on the gansynth subset from tensorflow_datasets, which requires a 4 line tweak to the code. If it's still not working for you on 1.13 I'll go ahead and make a PR to update the dataset loading to use tensorflow_datasets.

**Comment by jesseengel** [comment]
correction, it was trained on the full dataset downloaded from tfds, but with the filtering already in the gansynth code.

**Comment by shansiliu95** [comment]
> correction, it was trained on the full dataset downloaded from tfds, but with the filtering already in the gansynth code.

Maybe I should re-train using tfds instead of the tfrecord?

**Comment by shansiliu95** [comment]
Jesse, it would be great if you can provide some code on how to use tensorflow_datasets.

**Comment by jesseengel** [comment]
Sure thing, making a PR now. It would still be helpful if you can try training on 1.13 to so if it's a problem from 1.14-1.15

**Comment by shansiliu95** [comment]
Thanks! I am training on 1.13.

**Comment by jesseengel** [comment]
So, I forgot our internal PRs aren't visible in this repo (as they are in ddsp). Here's the code I'm trying now to reproduce my October result:

Download NSynth

```
import apache_beam as beam
import tensorflow_datasets as tfds

nsynth_builder = tfds.builder('nsynth')
download_config = tfds.download.DownloadConfig(
    beam_options=beam.options.pipeline_options.PipelineOptions())

nsynth_builder.download_and_prepare(download_config=download_config)
```

Changes to dataset


![Screen Shot 2020-05-28 at 4 07 11 PM](https://user-images.githubusercontent.com/1724681/83203004-7b0f6d00-a0fd-11ea-85b0-ebee2ce5876b.png)

**Comment by shansiliu95** [comment]
> nsynth_builder

Thanks for your code! One more question is how to integrate "nsynth_builder" into dataset.py?

**Comment by shansiliu95** [comment]
I have finished training with tf 1.13.1, the model still fails to generate good sounds

**Comment by jesseengel** [comment]
Okay, that helps narrow it down, so it sounds likely to be a dataset issue of some sort. Sorry for the delays (a lot going on at the moment) but I'm going to set my GCP instance to train on the dataset downloaded from https://magenta.tensorflow.org/datasets/nsynth to see if it reproduces your result

**Comment by shansiliu95** [comment]
Hey Jesse, any update?

**Comment by jesseengel** [comment]
Hi, sorry for the delay, in the middle of putting a paper together. I just checked and can confirm that it is an issue with the dataset. I reproduced the [not so great result](https://drive.google.com/file/d/1T71tpoOkyU1lbb-hxzXOKhNu5O6b3pCj/view?usp=sharing) by training on the NSynth dataset available on the website.

Adam highlighted that the issue is likely that the NSynth dataset from tensorflow_datasets has been shuffled by shards and is probably much more uniformly distributed, which helps with the GAN optimization. So, for now, training on the tensorflow_datasets version of the NSynth dataset should do the trick (you don't need to use the gansynth_subset since the gansynth code already filters the data [The model I trained on GCP that works used the full dataset as input, although the two should have the same result]).

**Comment by adarob** [comment]
Aside from the shuffling, the gansynth_subset actually uses a different
train/test split that should help with training. The original dataset
splits were created so that all pitch/velocities of each instrument were in
exactly one split. In the gansynth_subset, the same instrument may appear
in different splits but not with the same pitch/velocity.

**Comment by shansiliu95** [comment]
Problem solved! I re-train with tfds. Thanks for your help :)

**Comment by lonce** [comment]
@shansiliu95 - Hey, that's fantastic you were able to reproduce this data! 
What did you have to do to make this work?
Downloading the tfds nsynth dataset creates multiple tfrecords. Did you somehow combine them to provide the one tfrecord argument that gansynth_train.py expects, or did you make changes to the code in the repository to use tfds to load the data? 
(BTW, I am the one who submitted the previous issue you cited - #1607).
Many thanks.

**Comment by shansiliu95** [comment]
> @shansiliu95 - Hey, that's fantastic you were able to reproduce this data!
> What did you have to do to make this work?
> Downloading the tfds nsynth dataset creates multiple tfrecords. Did you somehow combine them to provide the one tfrecord argument that gansynth_train.py expects, or did you make changes to the code in the repository to use tfds to load the data?
> (BTW, I am the one who submitted the previous issue you cited - #1607).
> Many thanks.

Instead of using TFRecords, you can use this function in datasets.py

  def _get_dataset_from_tfds(self):
    nsynth_builder = tfds.builder('nsynth')
    download_config = tfds.download.DownloadConfig(
        beam_options=beam.options.pipeline_options.PipelineOptions())
    nsynth_builder.download_and_prepare(download_config=download_config)
    ds = nsynth_builder.as_dataset(split='train', shuffle_files=True)
    ds = ds.apply(contrib_data.shuffle_and_repeat(buffer_size=1000))
    return ds

**Comment by lonce** [comment]
Can you let me know what tensorflow version are you running?
I'm using 2.2.0 and tfds works fine, but after getting the dataset, the next line bombs: 
ds.map(_parse_nsynth, num_parallel_calls=4)
     TypeError: Expected any non-tensor type, got a tensor instead.
This happens in
tensorflow/python/data/ops/dataset_ops.py
so I suspect a version mismatch of some sort.

Thanks again,

**Comment by shansiliu95** [comment]
> @shansiliu95 - Hey, that's fantastic you were able to reproduce this data!
> What did you have to do to make this work?
> Downloading the tfds nsynth dataset creates multiple tfrecords. Did you somehow combine them to provide the one tfrecord argument that gansynth_train.py expects, or did you make changes to the code in the repository to use tfds to load the data?
> (BTW, I am the one who submitted the previous issue you cited - #1607).
> Many thanks.

I am using Tensorflow 1.15

**Comment by lonce** [comment]
I have not been able to reconstruct the compatible combination of legacy tensorflow, magenta, tfds, and nsynth versions in order to reproduce the results reported in the GANSynth paper. 

I thought my best hope might be to move to Tensorflow 2.2.0, the latest tdfs and magenta, and nsynth 2.3.2. I am still struggling to make the necessary changes to datasets.py to accommodate tdfs, new nsynth feature formats, etc. Humans reproduce faster than these results!

**Comment by lonce** [comment]
Yes! I have also been able to reproduce the results of the paper. Whew. 
I am using TF2, and tfds with the gansynth dataset. I had to make tweaks to dataset.py (for the new data format), and to data_helpers.py for version updating. Also, the hard-coded pitch_counts in datasets.py are incorrect for the gansynth data sets (the sum of the counts=81454 in the code, but nsynth/gansynth_subset has 60,788 in the train set and 886,755 in all). 
Thanks,
